### ISSUE TITLE
Implement PIN Auth

### ISSUE THREAD

**Issue by Razzeee** [fuzeman/trakt.py#21]
Token auth will soon be deprecated
See:

https://plus.google.com/u/0/109991097847241028373/posts/dCvyQnUJ5Ea?cfem=1

**Comment by fuzeman** [comment]
Should have this implemented in trakt.py today.

Might have some trouble using PIN authentication in Plex though :confused:

**Comment by fuzeman** [comment]
See https://github.com/fuzeman/trakt.py/blob/develop/examples/pin.py for an example on how to use PIN authentication.

Changes (in develop):
 - Added `Trakt['oauth'].pin_url()` method
 - Added "id" parameter to `Trakt.configuration.app()` method *(required when calling the pin_url() method)*
 - Added some extra checks to ensure `Trakt['oauth']` requests are sent with valid parameters

**Comment by Razzeee** [comment]
Damn, your fast!
Why the are you running into problems with plex?

**Comment by fuzeman** [comment]
Yeah, turned out PIN authentication was very simple to implement :smile:

Plex plugins have a very limited UI, not sure I'll be able to ask the user for the PIN easily... It might be technically possible but it still needs to be simple enough for users to understand.

**Comment by Razzeee** [comment]
I will have similar problems as I'm not sure what the settings panel is capable of. 
Let's hope you can solve the plex integration

**Comment by Razzeee** [comment]
What about refreshing the token? Do we need to implement this too?

**Comment by fuzeman** [comment]
The problem with PIN authentication (in Plex) is displaying the input dialog, it looks like only a few Plex clients actually display it properly...

Maybe we could get PIN authentication where the client displays a code that is then entered into the trakt.tv website, similar to how plex.tv authentication is implemented:
![](http://i.imgur.com/Jc6dccd.png)

---------------------------------------------------

I'll need to change ```Trakt['oauth'].token()``` to support the "refresh_token" parameter, will have this ready later today.

**Comment by Razzeee** [comment]
That's also how chromecast is doing this, might be worth to have a look at

**Comment by rectifyer** [comment]
We considered the Plex approach too, but it would be a custom implementation on our side. It would also require your app to do some sort of polling to know if the pin was entered.

The way we have PIN auth now is similar to how Twitter handles it. We were also just able to use standard OAuth which saves us from rewriting most of it. 

I'm still open to ideas based on the challenges you encounter. My main idea was the settings panel can accept the PIN and then on first run you do the OAuth exchange. If the settings panel is dynamic at all, then you could replace the PIN field indicating you have a connected account. That's what I had in my mind how this could work anyway.

**Comment by fuzeman** [comment]
Having the PIN entered in the settings dialog probably isn't too bad, displaying the PIN URL in the label is a bit strange but there isn't really any way around that.

@rectifyer will tokens generated from ```/auth/login``` continue to work (after it's deprecated)? or will they be revoked?

---------------------------------------

**Plex/Web**
![](https://i.imgur.com/tDoEVdP.png)

**Plex Home Theater**
![](https://i.imgur.com/0walxTM.png)

**Comment by rectifyer** [comment]
Yeah, I think that UI looks fine. It's not something people will really do more than once.
The token methods will still work as is, we just won't be giving access or making changes to them anymore. Existing apps that use it will continue working as they do now.
Also, PIN auth is only in staging right now. I'll push it live once we make sure there aren't any further changes. Probably tomorrow.





On Mon, Apr 6, 2015 at 7:33 PM -0700, "Dean Gardiner" <notifications@github.com> wrote:












Having the PIN entered in the settings dialog probably isn't too bad, displaying the PIN URL in the label is a bit strange but there isn't really any way around that.



@rectifyer will tokens generated from /auth/login continue to work (after it's deprecated)? or will they be revoked?





Plex/Web





Plex Home Theater

**Comment by fuzeman** [comment]
@Razzeee I've added the method ```Trakt['oauth'].token_refresh()```, look at the latest "pin.py" example on the develop branch for an example of token refreshing. *(the ```token()``` method has been renamed to ```token_exchange()```)*

Token refreshing could be handled inside trakt.py to simplify things though, will work on implementing this today/tomorrow.

**Comment by Razzeee** [comment]
Looking good. But yeah, I think it

**Comment by Razzeee** [comment]
Looking good. But yeah, I think it should handle it internally

**Comment by fuzeman** [comment]
@Razzeee automatic token refreshing is available on the develop branch now and ```examples/pin.py``` has been updated with more details.

Just make sure you update the OAuth configuration like:
```python
# Exchange PIN code with token
authorization = Trakt['oauth'].token_exchange(pin, 'urn:ietf:wg:oauth:2.0:oob')

with Trakt.configuration.oauth.from_response(authorization, refresh=True):
    # Send authenticated calls
    pass
```

and update the stored authorization details whenever the  ```oauth.token_refreshed``` event is fired:
```python
def on_token_refreshed(authorization):
    # Save authorization for future calls
    pass

Trakt.on('oauth.token_refreshed', on_token_refreshed)
```

----------------------------------

**Note:** Haven't finished implementing PIN authentication in the plex plugin yet, so there still could be some bugs.

**Comment by Razzeee** [comment]
@fuzeman why do I have to add refresh=True? Are there cases I don't want a refresh to happen? Shouldn't this be default?

**Comment by fuzeman** [comment]
@Razzeee if someone isn't handling tokens correctly (with the ```oauth.token_refreshed``` event), tokens could be refreshed inside every ```Trakt.configuration.oauth``` context.

I've decided to disable refreshing by default so people don't accidentally hit the trakt API with excessive requests.

**Comment by Razzeee** [comment]
So that's only if someone doesn't use the event? 
Can you give an "bad" example?

**Comment by Razzeee** [comment]
And if someone is using it wrong it's pretty likely that he will still have set the refresh to true?

**Comment by fuzeman** [comment]
An example of incorrect authorization/token handling:
```python
authorization = # <read token/authorization from storage>

with Trakt.configuration.oauth.from_response(authorization, refresh=True):
    Trakt['search'].query('Breaking Bad')

with Trakt.configuration.oauth.from_response(authorization, refresh=True):
    Trakt['search'].query('Game of Thrones')
```

As the ```authorization``` variable is never updated when the token refreshed, you will be just sending the old token (which is then refreshed again).

-----------------------------------------------------

Just to be clear, ```refresh=True``` is required if you want authorization tokens to be refreshed automatically.

If someone isn't aware of token refreshing, they probably won't have ```refresh=True```. So, when the authorization token expires a warning will be displayed instead.

**Comment by Razzeee** [comment]
Yeah, but I will have set this for each call. Wouldn't a global setting be much better?

**Comment by fuzeman** [comment]
All configuration parameters can have defaults set, just use ```Trakt.configuration.defaults```.

**Example:**

Set your defaults once:
```python
Trakt.configuration.defaults.oauth(
    refresh=True
)
```

Then you can omit the "refresh" parameter from your ```oauth.from_response()``` calls:
```python
with Trakt.configuration.oauth.from_response(authorization):
    # authenticated calls
```

**Note:** found a minor bug in the oauth configuration methods (when using defaults), update trakt.py (from the develop branch) for a fix.

**Comment by Razzeee** [comment]
Thank you!

**Comment by fuzeman** [comment]
PIN/OAuth changes have been released with v2.3.0
